baaas-decision-fleet-manager-api/src/main/java/org/kie/baaas/dfm/api/decisions/validators/DecisionNameValidator.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.kie.baaas.dfm.api.decisions.validators;
+
+import java.util.regex.Pattern;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+/**
+ * A validator validates the decision name in a DecisionRequest to ensure the length is between 1 and 253,
+ * the content is of lower case alphanumeric characters or '-', start with an alphabetic and end with
+ * alphanumeric characters.
+ *
+ * @see DecisionName
+ */
+public class DecisionNameValidator implements ConstraintValidator<DecisionName, String> {
+
+    final static String regExp = "[a-z]([-a-z0-9]*[a-z0-9])?";
```

Comment:
I'm not convinced we should force the user to use K8S compatible names for their Decision. We need to remember who is going to be typically responsible for naming a Decision and that will probably be the Decision Developer persona, who may not know about or understand Kubernetes. We also shouldn't allow our internal choice of implementation (e.g Kubernetes) control what the user can do via our API. So for example, I think the following should be acceptable Decision names:

- Payroll - Wage Increase Decision
- 1st Of Month Loan Decision
- Rob's First Decision

I believe those are the types of name that the Decision Developer will work with. It's up to us to handle that and convert it appropriately. 

It _probably_ makes sense to perform a case-insensitive check to see if there is a Decision with the same name so that the following are seen as equivalent:

- My first Decision
- MY FIRST DECISION

That I believe will solve the majority of the problem here. There may be some additional consideration about what to do if the name starts with a number or contains punctuation e.g `'`, but perhaps we simply add a reserved prefix to the name in the `spec` of the `DecisionRequest` e.g:

- 1st Decision --> `num-1st-decision`

So for me it's more about how we convert the name into the `spec` of the `DecisionRequest` CR. 

@r00ta / @myeung18 : WDYT?

Reply:
Hi @robpblake , yes I agree that those should be valid decision names. 

What if we use the `decisionId` as resource name instead of the `name`?

Reply:
I overlooked the UI part. That makes sense to have names like that.


The current sanitizer method already does almost everything except the leading numeric conversion.  https://github.com/kiegroup/baaas-decision-fleet-manager/blob/1aa3960f3e8881817c571dff02808d7a549839bb/baaas-decision-fleet-manager/src/main/java/org/kie/baaas/dfm/app/dfs/client/DefaultDecisionFleetShardClient.java#L89. we just need to add one more check and append that 'num-' in the front for leading num.

Reply:
@r00ta I'm not sure of the impact of using the `id` of the Decision, I wonder if it will mean we just end up generating invalid resource names in the Fleet Shard. `id` is always a UUID, so it maybe that we try to generate names for things like the `PipelineRun` that are not valid 🤔